src/analyses/weaklyRelationalPointerAnalysis.ml

```diff
@@ -0,0 +1,197 @@
+    (* Indefinite assignment *)
+    | s, lterm, (None, _) -> D.remove_may_equal_terms ask s lterm t
+    (* Definite assignment *)
+    | s, lterm, (Some term, Some offset) ->
+      let dummy_var = MayBeEqual.dummy_var (typeOfLval lval) in
+      if M.tracing then M.trace "wrpointer-assign" "assigning: var: %s; expr: %s + %s. \nTo_cil: lval: %a; expr: %a\n" (T.show lterm) (T.show term) (Z.to_string offset) d_exp (T.to_cil lterm) d_exp (T.to_cil term);
+      t |> meet_conjs_opt [Equal (dummy_var, term, offset)] |>
+      D.remove_may_equal_terms ask s lterm |>
+      meet_conjs_opt [Equal (lterm, dummy_var, Z.zero)] |>
+      D.remove_terms_containing_variable @@ MayBeEqual.dummy_varinfo (typeOfLval lval)
+    | exception (T.UnsupportedCilExpression _) -> D.top ()
+    (* the assigned variables couldn't be parsed, so we don't know which addresses were written to.
+       We have to forget all the information we had.
+       This should almost never happen.
+       Except if the left hand side is an abstract type, then we don't know the size of the lvalue. *)
+    | _ -> D.top ()
+
+  let assign ctx lval expr =
+    let res = assign_lval ctx.local (ask_of_ctx ctx) lval expr in
+    if M.tracing then M.trace "wrpointer-assign" "ASSIGN: var: %a; expr: %a; result: %s. UF: %s\n" d_lval lval d_plainexp expr (D.show res) (Option.map_default (fun r -> TUF.show_uf r.uf) "" res); res
+
+  let branch ctx e pos =
+    let props = T.prop_of_cil (ask_of_ctx ctx) e pos in
+    let valid_props = T.filter_valid_pointers props in
+    let res = meet_conjs_opt valid_props ctx.local in
+    if D.is_bot res then raise Deadcode;
+    if M.tracing then M.trace "wrpointer" "BRANCH:\n Actual equality: %a; pos: %b; valid_prop_list: %s\n"
+        d_exp e pos (show_conj valid_props);
+    res
+
+  let body ctx f = ctx.local (*DONE*)
+
+  let assign_return ask t return_var expr =
+    (* the return value is not stored on the heap, therefore we don't need to remove any terms *)
+    match T.of_cil ask expr with
+    | (Some term, Some offset) -> meet_conjs_opt [Equal (return_var, term, offset)] t
+    | _ -> t
+
+  let return ctx exp_opt f =
+    let res = match exp_opt with
+      | Some e ->
+        assign_return (ask_of_ctx ctx) ctx.local (MayBeEqual.return_var (typeOf e)) e
+      | None -> ctx.local
+    in if M.tracing then M.trace "wrpointer-function" "RETURN: exp_opt: %a; state: %s; result: %s\n" d_exp (BatOption.default (MayBeEqual.dummy_lval (TVoid [])) exp_opt) (D.show ctx.local) (D.show res);res
+
+  let special ctx var_opt v exprs  =
+    let desc = LibraryFunctions.find v in
+    match desc.special exprs, v.vname with
+    | Assert { exp; refine; _ }, _ -> if not refine then
+        ctx.local
+      else
+        branch ctx exp true
+    | _, _ -> ctx.local
+
+  let duplicated_variable var = { var with vid = - var.vid - 4; vname = "wrpointer__" ^ var.vname ^ "'" }
```

Comment:
This is fine for now, but we have more sophisticated facilities for maintaining copies of variables for such purposes (see the privatizations in apron).